### PR TITLE
Introduce config option for default generic network options of newly created networks

### LINKED ISSUE #23938: User-specified network driver option defaults
It is possible to configure network drivers at network create time:

```
docker network create -d overlay -o "com.docker.network.mtu=9001" foobar
```

In some cases, you need to change an option for _all_ networks created on a system. Right now, the way to accomplish that will be to change _every_ compose file, Makefile, bash script, or other deploy mechanism to accommodate this host-wide requirement. This is not a scalable solution.

There should be a ui/ux that would allow a user to set defaults system-wide: "all overlay networks will need to have this value for that option".

Implementing an alternate driver that has the needed behavior is not a viable workaround. It requires the same amount of effort to change _all_ compose/make/bash/deploy scripts to use the alternate driver, so this won't address the issue.

### LINKED ISSUE #34981: Docker daemon MTU not used when creating a new bridge network
**Description**
I configure the `/etc/docker/daemon.json` with the "mtu" parameter to 1450.
Default bridge is ok with mtu=1450.
But when a new bridge network is created (either `docker network create `or `docker-compose`), the mtu parameter is false. It's always 1500.

**Steps to reproduce the issue:**
1. Add `"mtu": 1450 `to `/etc/docker/daemon.json`
2. Restart docker daemon
3. Create a new bridge : `docker network create my-net`

**Describe the results you received:**
my-net MTU = 1500 

**Describe the results you expected:**
my-net MTU = 1450

**Output of `docker version`:**

```
Client:
 Version:      17.06.2-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   cec0b72
 Built:        Tue Sep  5 20:00:17 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.2-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   cec0b72
 Built:        Tue Sep  5 19:59:11 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 4
 Running: 1
 Paused: 0
 Stopped: 3
Images: 41
Server Version: 17.06.2-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/100000.100000/aufs
 Backing Filesystem: extfs
 Dirs: 163
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 6e23458c129b551d5c9871e5174f6b1b7f6d1170
runc version: 810190ceaa507aa2727d7ae6f4790c76ec150bd2
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
 userns
Kernel Version: 4.4.0-93-generic
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 15.67GiB
Name: test-vm-04
ID: OZSP:AWLJ:HAPO:CZD6:NNLX:YDIT:LGWR:RMEI:P4SK:SBY4:2RAT:5YIK
Docker Root Dir: /var/lib/docker/100000.100000
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: true
```

### PR DESCRIPTION
**Pull request by dajudge**
**- What I did**
I introduced a new command line parameter `--default-network-opt` to configure defaults for generic options for newly created networks (e.g. MTU).

For example you could configure the MTU for newly created bridge networks by using the following command line:

```shell
dockerd --default-network-opt=bridge=com.docker.network.driver.mtu=1234
```
**Precendence rules:**
* Should there be a configuration option specified as parameter of the network creation then this will always take precedence over the default provided via the cmdline

**- How I did it**
* I introduced a new cmdline param `--default-network-opts`. Its values are stored in a newly created `DefaultNetworkOpts` in the struct `NetworkConfig` found in the file `src/github.com/docker/docker/daemon/config/config.go`
* `createNetwork()` in `src/github.com/docker/docker/daemon/network.go` picks up the values from `NetworkConfig.DefaultNetworkOpts` and adds them to the network options provided when creating a new network based on the type of the created network.
  * Options provided as parameters for the network creation take precedence over the defaults

**- How to verify it**
* Run `dockerd --default-network-opt=bridge=com.docker.network.driver.mtu=1234` (or just `dockerd` without this PR)
* Create a new network via `docker network create mynet`
* Check the MTU of the newly created network with `docker run --rm --net mynet -it ubuntu  cat /sys/class/net/eth0/mtu`
  * Without this PR the MTU of the network should default to 1500
  * With this PR the MTU of the network is set to 1234

**- Description for the changelog**
Introduce the `--default-network-opt` configuration option for newly created networks.

- fixes #34981
- fixes https://github.com/moby/moby/issues/23938

### CODE REVIEW

File: daemon/config/config.go
Comment (bsideup) [inline daemon/config/config.go:124]: Overlay networks could/should use this parameter too

```suggestion
	// Mtu allows to specify the MTU used for creating custom networks if nothing else is specified upon creation
```

**Comment by olljanat** [comment]
@dajudge can you squash this to single commit and rebase with master branch?

I did quick testing with this and it looks to be effective for all new networks, including ingress and all custom overlay networks as long it is set to manager node which means that it would help also with this case https://github.com/moby/libnetwork/issues/2661 Will do testing with Windows also after rebase is done (now I was not able to build on Windows).

@thaJeztah @cpuguy83 is this something which can be considered to be included to Moby and maybe also backported to 22.06 ? It is kind of breaking change but because most of the environments are anyway using only default MTU and those who changes it will do it on purpose it would make sense to simplify their life with this.

**Comment by dajudge** [comment]
Hi @olljanat - thanks for your feedback!

If the potentially breaking nature of the patch is a problem it'd also be possible to introduce a new cmdline arg instead of reusing `--mtu` - just let me know what makes more sense.

I'll gladly take care of rebasing in the next few days if that helps getting this PR merged! :+1:

**Comment by dajudge** [comment]
> can you squash this to single commit and rebase with master branch?

@olljanat done :+1:

File: daemon/config/config.go
Comment (olljanat) [inline daemon/config/config.go:292]: Moving this looks to be breaking this test
https://github.com/moby/moby/blob/f34567bf41ad65102a8b1f05496dc92b500a3056/daemon/config/config_test.go#L255-L258
and this test
https://github.com/moby/moby/blob/f34567bf41ad65102a8b1f05496dc92b500a3056/daemon/config/config_test.go#L411-L414

so probably it should be better to add new flag like `--default-mtu` like there looks to be couple of existing parameters like those already and add test cases for that (at least MTU with and without that value) as maintainer always like to see those.

Comment (dajudge) [reply daemon/config/config.go]: I went for `--network-default-mtu` - WDYT?

Comment (olljanat) [reply daemon/config/config.go]: Fine for me.

**Comment by dajudge** [comment]
I'd love to get some feedback on this PR from you folks (@olljanat, @cpuguy83, @thaJeztah) to learn what could help bringing it forward.

Do you maybe have some ideas regarding tests that could be added? What do you think about the overall concept of introducing a new cmdline flag for this?

I can also help backporting if it is considered valuable to have it in 22.06 as @olljanat suggested.

Your input is greatly appreciated!

**Comment by olljanat** [comment]
@dajudge purpose of tests is proof that feature works as expected and more importantly make sure that it does not go broken as part of future code changes. Because there have been already that earlier MTU setting there most probably is some tests which can be extended or alternatively you can write new one(s). Looks folder `integration` about how to build those and maybe create something which run dockerd with this settings enabled, create custom bridge network, start container and check inside of it that MTU value is correct.

**Comment by cpuguy83** [comment]
New flag seems ok to me.
Marking this for code review.

Just need a test and I think we can move forward.

**Comment by dajudge** [comment]
Added test case to verify behavior in the following cases:
* Defaults to MTU=1500 when no cmdline args are passed
* Applies MTU when provided via cmdline

Each case is verified by creating a new network, starting a container attached to that network and then checking the MTU of the eth0 interface in the container.

**Comment by sum-catnip** [comment]
Not to be annoying but is there any plans of integrating this feature yet? Id love to use it!

**Comment by dajudge** [comment]
Resolved some merge conflicts.

**Review by sum-catnip — APPROVED** [review]
wait am i able to approve this??

**Comment by sum-catnip** [comment]
uuuuhm

**Review by cpuguy83 — COMMENTED** [review]
This seems ok, though I'd expect that such a default would be fairly driver specific.

**Comment by s4ke** [comment]
+1 on this. On our cloud provider when using Ubuntu 22.04 something has changed that caused the default mtu to break overlay networks. For now we can live with setting the mtu in our stack files but a default would help newcomers.

**Comment by sum-catnip** [comment]
in my setup i have to put
```yml
networks:
  default:
    driver_opts:
      com.docker.network.driver.mtu: 1420
```

into every single service im hosting.
It's not only incredibly annoying but also not portable at all.

can we please have this feature merged?
ill even take it as a late christmas gift so you dont have to get me anything

**Comment by cpuguy83** [comment]
Thinking about this a little bit:

1. If we pass a default value to drivers, I think the option should be driver specific (so something like `--network-default-mtu=<driver>=<value>`.
2. It might be nice to introduce a new standard option for max MTU which the driver itself can then go and subtract it's own overhead from. This allows the driver to do its own default rather than pushing a default config down from docker.

This is said with the assumption that drivers are responsible for setting MTU and not libnetwork... I haven't taken the time to dig into the code and am currently on PTO.

**Comment by cpuguy83** [comment]
So yes, as suspected this should all be handled by the driver so we either need to namespace the options for each driver or pass down an option to drivers for max MTU.

Max MTU is of course a pretty big change since every driver would need to be changed to support it.

**Comment by dajudge** [comment]
Hi @cpuguy83,

many thanks for your input!

Disclaimer: I'm far from being an expert on the nitty gritty networking layer details that are involved here, so please bear with me if this idea would not work for some reason I'm not taking into account.

What about leaving the parameter as-is but taking a driver-specific overhead into account before applying it via 
```
genericData[netlabel.DriverMTU] = fmt.Sprintf("%d", c.cfg.DefaultMTU)
```

My thinking here is: 

https://github.com/moby/moby/blob/eaa7b49bd6d968b69c9c8b2d34746c4fa43500d9/libnetwork/controller.go#L669 already fetches driver-specific information. So how about extending the `driverapi.Capability` struct to also include a driver-specific value for network overhead. If such a constant value could be determined for every driver then one could simply subtract that value from `c.cfg.DefaultMTU` before setting it as the `DriverMTU`, e.g. as such:

```
genericData[netlabel.DriverMTU] = fmt.Sprintf("%d", c.cfg.DefaultMTU - cap.PacketOverhead)
```

WDYT? Would it make sense to go down such a road?

**Comment by cpuguy83** [comment]
cc @corhere 

@dajudge So you are saying to keep a global default and then a new driver capability to return packet overhead?
Perhaps that could work. We could just take this change in as is.

**Comment by corhere** [comment]
The drivers are responsible for setting MTU. And the only drivers today which do so are bridge and overlay. I much prefer the idea of introducing a new "Max MTU" option for drivers to subtract their overhead from over the capability idea. The packet overhead might not necessarily be a per-driver constant: per-network options could influence the overhead, for instance.

As for this PR, I have to wonder whether we should be adding a special case daemon config option just for one particular network config option. What about `com.docker.network.enable_ipv6`, `com.docker.network.container_iface_prefix`, etc.? Should we have a bespoke daemon flag for every one? I'd be more inclined to have the daemon configuration be more open-ended by allowing defaults to be set for any per-network option. E.g.:

```
dockerd \
  --default-network-opt=overlay=com.docker.network.driver.mtu=1420 \
  --default-network-opt=*=com.docker.network.container_iface_prefix=foo-
```

```json
{
  "default-network-opts": {
    "*": {"com.docker.network.container_iface_prefix": "foo-"},
    "overlay": {"com.docker.network.driver.mtu": "1420"},
    "bridge": {"com.docker.network.driver.mtu": "1450"}
  }
}
```

**Comment by dajudge** [comment]
@corhere @cpuguy83 Thanks for having a closer look!

So I picked up your idea of a more generic `--default-network-opt` param and hacked together an experiment to see what it'd feel like.

For starters I simplified the option parsing by just using JSON on the cmdline like this:
```
--default-network-opt='bridge={"com.docker.network.driver.mtu":"1234"}'
```

I like how things turned out here and would love to get some input from you folks on this before I run too far in an unfavorable direction. WDYT?

**Comment by corhere** [comment]
I like how things are turning out, too.

JSON on the command line is not terribly ergonomic to write and needs quoting/escaping, but then again so does the use of the `*` in my strawman. I think we can punt bikeshedding the command-line flag's syntax until later. I am more interested in codifying the semantics of how the defaults are combined with the provided options when creating a new network. The current implementation appears to perform a merge of the option sets in decreasing order of precedence: provided options, `*` defaults, networkType defaults.
- I personally agree with merge semantics. It's what I was imagining when writing my earlier comment. I want to call it out so everyone is aware and has their chance to voice dissenting opinions.
- networkType defaults should have a higher precedence than `*` defaults, such that more-specific configuration always overrides less-specific. But that's probably what you meant.
- While merging semantics allow a default option to be overridden by changing its value, the option itself will still be set. Does there need to be an affordance to also _unset_ a default option when creating a new network? Setting `com.docker.network.container_iface_prefix` to the empty string will have the same result as the option not being set at all, but is that universally true for all network options?
  - My opinion on this is YAGNI, for the record. I don't think it is a requirement for default network options to be useful, and could be added later if there is user demand for it.
- Should the default options be applied to the special-case default bridge network? I'm thinking it shouldn't, as it is created implicitly and is special in lots of other ways. We would be able to summarize the semantics of `--default-network-opt` as being applied to any new networks created by users.

As for the implementation, I would prefer if it was kept out of libnetwork. It is a daemon config option interacting with a daemon API, therefore I think the daemon should be responsible for implementing the defaults-merging semantics. `(*Daemon).createNetwork` is the single place where all code paths to create networks converge (aside from the default `none`, `host`, `bridge` networks) and so defaults could be applied [by merging them into `create.Options`.](https://github.com/moby/moby/blob/3ba527d82af53cf752aa6ec39daac14626c41b0f/daemon/network.go#L320)

**Comment by cpuguy83** [comment]
My only gripe is the wildcard itself.
I'd replace that with a string like `com.mobyproject.network.defaults`

File: daemon/network.go
Comment (corhere) [inline daemon/network.go:318]: Maps are reference types so `networkOptions` is just an alias. The `create.Options` map is still mutated in-place. Either ditch the misleading alias or deep-copy the map.

File: daemon/network.go
Comment (corhere) [inline daemon/network.go:322]:
```suggestion
				logrus.Debugf("Applying default option %v=%v to network %v", k, v, id)
```
(There's probably a structured log field for network IDs but I can't think of what the field name might be.)

**Review by corhere — COMMENTED** [review]
The PR title and description could use an update given the expanded scope. I wholeheartedly approve of all the semantics as currently implemented.
- Only applies to user-defined networks, on create
- Precedence: per-network options override driver defaults override wildcard defaults
- No affordances for unsetting a default option

**Comment by dajudge** [comment]
> The PR title and description could use an update given the expanded scope.

@corhere I updated title and desc. Please let me know if you feel that there's room for improvement!

**Comment by dajudge** [comment]
I changed the cmdline arg parsing to mirror what @corhere originally suggested and added some tests around it. PR desc is updated accordingly.

WDYT?

**Comment by corhere** [comment]
Lint errors:
```
opts/opts.go:186:5: S1003: should use !strings.Contains(value, "=") instead (gosimple)
	if strings.Index(value, "=") == -1 {
	   ^
opts/opts.go:190:5: S1003: should use !strings.Contains(rv, "=") instead (gosimple)
	if strings.Index(rv, "=") == -1 {
	   ^
```

**Comment by corhere** [comment]
We could not come to a consensus on the wildcard defaults feature, including whether it is even desired. Aside from that the maintainers are happy with the design. Remove that feature from this PR and it'll be good to go, design-wise. Wildcard defaults could be reintroduced in a follow-up PR where we could bikeshed all the nuances of the design without holding up this PR any longer.

(For future reference, we all agree that `*` is not the right choice of wildcard because of shell-escaping issues, `default` is off the table because it's already a very overloaded keyword and has special meaning when creating networks because Windows, and I suggested `_all`. @thaJeztah and @neersighted are concerned that users could misuse wildcard defaults as a too-big hammer to set options which don't make sense to set to the same value for all network drivers, and fill up our issue tracker with garbage when they shoot themselves in the foot.)

**Comment by thaJeztah** [comment]
Thanks for summarizing the discussion, @corhere

For the wildcard; this was a "No is temporary, Yes is forever" (https://github.com/moby/moby/blame/v1.0.0/hack/PRINCIPLES.md#L13) decision. We are considering it, but it may need further discussion, and _not_ implementing it (for now) keeps our options open (whereas the "reverse" is a lot more complicated).

I should also add that we were discussing if it would potentially be possible (in future) to "soft-deprecate" (hide) the `--mtu` flag (which sets the MTU for the default bridge network), and to combine it with the feature implemented in this PR. That also needs further investigating / discussion, so needs to be revisited.

File: daemon/network.go
Comment (thaJeztah) [inline daemon/network.go:324]: nit: should we consider including the driver, and using structured logs for this?

```go
logrus.WithFields(logrus.Fields{"driver": driver, "network": id, k: v}).Debug("Applying default option")
```

**Review by thaJeztah — COMMENTED** [review]
changes and design LGTM (thanks!)

I see there's 13 commits in this PR now; I think some (or perhaps all) of them can be squashed?

File: daemon/network.go
Comment (dajudge) [reply daemon/network.go]: :heavy_check_mark:

**Comment by dajudge** [comment]
> I see there's 13 commits in this PR now; I think some (or perhaps all) of them can be squashed?

@thaJeztah squashed into 1 :+1:

**Comment by s4ke** [comment]
Thanks @dajudge for picking up this feature btw. This will make managing clusters a lot easier.

File: integration/network/network_test.go
Comment (corhere) [inline integration/network/network_test.go:223]:
```suggestion
			assert.Check(t, is.Contains(result.Combined(), fmt.Sprintf(" mtu %d ", tc.mtu)), "Network MTU should have been set to %d", tc.mtu)
```

**Review by corhere — APPROVED** [review]
LGTM!

**Comment by thaJeztah** [comment]
>  This will make managing clusters a lot easier.

only thing we need to look at; so it currently means that which manager you access will determine the defaults; do we want the defaults to be configured at the "swarm" level?

**Comment by neersighted** [comment]
I think for now this makes sense, all node configuration is node-local. We can consider designs for pushing 'config' (likely starting with plugins, but maybe some daemon.json fields/flags as well) in a separate step, I think.

**Comment by s4ke** [comment]
> > This will make managing clusters a lot easier.
> 
> only thing we need to look at; so it currently means that which manager you access will determine the defaults; do we want the defaults to be configured at the "swarm" level?

I can only speak for our usecase where we manage clusters via ansible - i think the current approach is fine. It might be nice and less surprising to have this as a config in the swarm, but yeah thats it, nice. 🙂

File: integration/network/network_test.go
Comment (dajudge) [reply integration/network/network_test.go]: lol thanks!

**Comment by thaJeztah** [comment]
Yes, maybe it's ok for a start, and maybe they're complimentary. We should have a look to verify the behavior (also for the documentation); thinking of the following situations;

- switching to another manager results in different results (if these are used when _creating_ the network's configuration)
- if these were _not_ used when _creating_ the network, but when "deploying" the network (swarm creates networks on each host when a task lands on them); are there options that can/should (technically?) differ between nodes?

If there's limitations (or things that the user should be aware of, such as "depends on the manager node you're connecting to") to the feature, then documenting those is fine. But also; if we need to keep future options open, we should have in mind if that's still possible (what would the behavior be if we added "swarm-scoped options later"?)

**Review by neersighted — APPROVED** [review]

File: daemon/network.go
Comment (corhere) [inline daemon/network.go:329]: Defaults should only be applied to networks where `create.ConfigFrom == nil` as [`ConfigFrom` cannot be used in combination with driver options.](https://github.com/moby/moby/blob/0021339b9299c5f1596bdf772a42ce346f16f060/libnetwork/network.go#L382-L415)

**Review by corhere — CHANGES_REQUESTED** [review]
Configuration code is deceptively tricky to get right; libnetwork configuration doubly so. Apologies for not flagging this gotcha earlier.

File: daemon/network.go
Comment (dajudge) [reply daemon/network.go]: Hope I got that right! :+1:

**Comment by dajudge** [comment]
FYI squashed the commits again.

**Review by neersighted — APPROVED** [review]

**Review by corhere — APPROVED** [review]

**Comment by neersighted** [comment]
@s4ke Could you go into more detail about what setting you control with Ansible? I'm not sure this PR really directly affects Swarm with the current design; certainly it doesn't seem to limit us or bear and special concerns WRT Swarm. I think any possible interactions with Swarm make more sense to consider in some of the follow-ups that have been discussed.

**Comment by dperny** [comment]
We have discussed in incredible detail how this affects Swarm and our determination is that we can merge this PR, from a Swarm standpoint.

We still do not *exactly* know how this affects Swarm, to be clear, we just think you can merge the PR anyway.

**Comment by neersighted** [comment]
(To be clear, how defaults end up in Swarm-created networks is something of a convoluted process; but the takeaway was that it really doesn't matter as to break overlay networks one must set the defaults for the overlay driver. For the bridges, they are configured and persisted node-local already, so this really doesn't change anything)

**Comment by s4ke** [comment]
> Could you go into more detail about what setting you control with Ansible? I'm not sure this PR really directly affects Swarm with the current design; certainly it doesn't seem to limit us or bear and special concerns WRT Swarm. I think any possible interactions with Swarm make more sense to consider in some of the follow-ups that have been discussed.

What we are doing right now is to set the following in our stack files because on Ubuntu 22.04 we faced issues if this was not there.

```
  traefik-public:
    driver: overlay
    attachable: true
    name: traefik-public
    driver_opts:
      encrypted: ""
      com.docker.network.driver.mtu: "1350"
```

What I was hoping with this PR is to have our ansible tooling set up the `/etc/docker/daemon.json` to set the defaults on all nodes in the swarm so that we don't have to set the `driver_opts` anymore.

**Comment by s4ke** [comment]
Should this be backported to 23.0?

**Comment by neersighted** [comment]
I do not believe so; this is a new feature that does not enhance the maintainability of the 23.0 branch, and does not address a significant defect in the branch. It's certainly something that is nice to have, but I don't think it's a good idea to introduce a net-new feature in a patch release.

**Comment by thaJeztah** [comment]
@richardg867 what version of docker are you running?

**Comment by corhere** [comment]
@richardg867 the `daemon.json` option is spelled `"default-network-opts"` – note the plural.

**Comment by thaJeztah** [comment]
Good call; also just realized we need to update the docs in https://github.com/docker/cli/blob/master/docs/reference/commandline/dockerd.md

**Comment by richardg867** [comment]
> @richardg867 what version of docker are you running?

24.0.0-beta.1

> @richardg867 the daemon.json option is spelled "default-network-opts" – note the plural.

After correcting the name, it now works for bridge networks, but overlay networks cause an unknown option error, even though they *do* work if I specify the option on the dockerd command line instead of `daemon.json`.

```
# cat daemon.json
{"default-network-opts": {"bridge": {"com.docker.network.driver.mtu": "1434"}}}```

# journalctl -xe
[...] Started Docker Application Container Engine. [...]
```

```
# cat daemon.json
{"default-network-opts": {"overlay": {"com.docker.network.driver.mtu": "1434"}}}```

# journalctl -xe
[...] unable to configure the Docker daemon with file /etc/docker/daemon.json: the following directives don't match any configuration option: overlay [...]
```

**Comment by richardg867** [comment]
Just found another interesting issue: default settings are not applied to networks created in swarm mode.

```
# docker network create --driver overlay test_net_please_ignore
bfbba07887feb61b818c67a0ad7f3c481a0991ed8bf722f0cfea33da6976c6dc

# docker network inspect bfbba07887feb61b818c67a0ad7f3c481a0991ed8bf722f0cfea33da6976c6dc
[...]
        "Options": {
            "com.docker.network.driver.mtu": "1434"
        },
[...]

# docker network rm bfbba07887feb61b818c67a0ad7f3c481a0991ed8bf722f0cfea33da6976c6dc
bfbba07887feb61b818c67a0ad7f3c481a0991ed8bf722f0cfea33da6976c6dc

# docker swarm init
Swarm initialized: current node [...]

# docker network create --driver overlay test_net_please_ignore
hjch5eyuxfspk1drvtgnnkz9r

# docker network inspect hjch5eyuxfspk1drvtgnnkz9r
[...]
        "Options": {
            "com.docker.network.driver.overlay.vxlanid_list": "4097"
        },
[...]
```

**Comment by corhere** [comment]
@richardg867 try inspecting the network while a container is attached to it.

**Comment by twatts-kcl** [comment]
Thanks very much for this patch - we use OpenStack so it's very useful.

Question: can it be expressed as a daemon.json config? I tried, but whilst this works:

`/usr/bin/dockerd -H fd:// --containerd=/run/containerd/containerd.sock --default-network-opt=bridge=com.docker.network.driver.mtu=1442`

This is accepted as valid syntax, but does not change the MTU (stays at default 1500 on creating user bridges):

`    "default-network-opt": 
    {
      "bridge": 
        {
          "com.docker.network.driver.mtu": 1442
        }
    },`
    
Yes, I can patch a config change into systemd for the docker service, but it would be nice if all I have is an error in my daemon.json which I could fix :)

Many thanks, Tim

**Comment by cpuguy83** [comment]
If the bridge already exists it doesn't change anything.
You'll have to stop docker, delete the bridge, then docker will recreate it.

**Comment by twatts-kcl** [comment]
> If the bridge already exists it doesn't change anything. You'll have to stop docker, delete the bridge, then docker will recreate it.

Hi, Thanks for taking the time to reply - I already did that. I've tried it both ways:

When I put --default-network-opt=bridge=com.docker.network.driver.mtu=1442 into the systemd docker.service file, reload systemd and restart docker, bringing up containers does teh right thing - new bridges are configured with MTU 1442.

When I remove that option from systemd, reload systemd, add the json config above into /etc/docker/daemon.json and restart docker: _The daemon restarts without error (which it doesn't do if I make a typo in that particular bit of the config) - But it seems to ignore the setting._

**I assumed that all dockerd command line options will mape to daemon.json settings?**

- So either they don't, in which case I will have to patch the systemd service config;
- Or they do, but I have not expressed the map correctly???

**Comment by twatts-kcl** [comment]
OK - Got it!

There is a disparity in the daemon.json and the command line options:

This works:

` "default-network-opts": {
      "bridge": {
        "com.docker.network.driver.mtu": "1442"
      }
},`

Notice the extra s on -opts.

I found this out with a dive into the source code:

config.go: `	flags.Var(opts.NewNamedMapMapOpts("default-network-opts", conf.DefaultNetworkOpts, nil), "default-network-opt", "Default network options")`

and

config_linux_test.go:

`
  "default-network-opts": {
    "overlay": {
      "com.docker.network.driver.mtu": "1337"
    }
  }
`

**Comment by neersighted** [comment]
The plural has always been how these types of options (that can be specified multiple times on the command line) have worked; cc @dvdksn to see if we can call this out louder for the daemon.json docs.

**Comment by twatts-kcl** [comment]
> The plural has always been how these types of options (that can be specified multiple times on the command line) have worked; cc @dvdksn to see if we can call this out louder for the daemon.json docs.

Maybe it's me, but I really struggled finding good daemon.json docs and examples.

It was actually the example in the config_linux_test.go that alerted me.

**Comment by thaJeztah** [comment]
Looks like docs were added in this PR;

- https://github.com/docker/cli/pull/4635

(published at https://docs.docker.com/reference/cli/dockerd/#default-network-options)

so removing the `docs/revisit` label

**Pull request merged**
